### PR TITLE
Enable vulnerability scanning of environment images

### PR DESCRIPTION
**Pull request by lykelly19**

### CODE REVIEW

File: scripts/azureml-assets/azureml/assets/environment/build.py [hunk 1]
```diff
@@ -62,6 +62,11 @@ def create_acr_task(image_name: str,
             'id': 'test',
             'cmd': f"$Registry/{image_name} {test_command}"
         })
+        task['steps'].append({
+            'id': 'scan',
+            'cmd': "wget https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb && dpkg -i trivy_0.42.1_Linux-64bit.deb && trivy fs --scanners vuln /",
+            'ignoreErrors': True
```
Comment (elesel) [inline scripts/azureml-assets/azureml/assets/environment/build.py:68]: I think you'll need the name of the image to run the command on and let's be explicit about the bool we pass for `ignoreErrors`.
```suggestion
            'cmd': f"$Registry/{image_name} wget https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb && dpkg -i trivy_0.42.1_Linux-64bit.deb && trivy fs --scanners vuln /",
            'ignoreErrors': 'true'
```

Comment (elesel) [reply scripts/azureml-assets/azureml/assets/environment/build.py:68]: You'll probably need to break this over multiple lines to make flake8 happy.

File: scripts/azureml-assets/azureml/assets/environment/build.py [hunk 2]
```diff
@@ -62,6 +62,12 @@ def create_acr_task(image_name: str,
             'id': 'test',
             'cmd': f"$Registry/{image_name} {test_command}"
         })
+        task['steps'].append({
+            'id': 'scan',
+            'cmd': f"$Registry/{image_name} wget https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb && \
+                    dpkg -i trivy_0.42.1_Linux-64bit.deb && trivy fs --scanners vuln /",
+            'ignoreErrors': True
```
Comment (elesel) [inline scripts/azureml-assets/azureml/assets/environment/build.py:69]: While this mostly works, it introduces a bunch of whitespace into the string. What do you think of this?
```suggestion
            'cmd': (
                       f"$Registry/{image_name} "
                       "wget https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb && "
                       "dpkg -i trivy_0.42.1_Linux-64bit.deb && "
                       "trivy fs --scanners vuln /"
                   ),
            'ignoreErrors': True
```

Comment (lykelly19) [reply scripts/azureml-assets/azureml/assets/environment/build.py:69]: That sounds better, thank you!

File: .github/workflows/environments-ci.yaml [hunk 3]
```diff
@@ -86,7 +86,7 @@ jobs:
 
       - name: Build and test changed images
         id: build-changed-images
-        run: python -u $scripts_environment_dir/build.py -i "${{ github.event.inputs.asset_dirs || env.default_asset_dirs }}" -a $asset_config_filename -o '${{ runner.temp }}'/$built_dir -l '${{ runner.temp }}'/$build_logs_artifact_name -P -g $resource_group -r $container_registry -T 'python -V' -c "${{ steps.changed-files.outputs.all_modified_files }}"
+        run: python -u $scripts_environment_dir/build.py -i "${{ github.event.inputs.asset_dirs || env.default_asset_dirs }}" -a $asset_config_filename -o '${{ runner.temp }}'/$built_dir -l '${{ runner.temp }}'/$build_logs_artifact_name -P -g $resource_group -r $container_registry -T 'python -V' -c "${{ steps.changed-files.outputs.all_modified_files }} -U 'https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb'"
```
Comment (elesel) [inline .github/workflows/environments-ci.yaml:89]: The quotes for `-c` are mismatched:
```suggestion
        run: python -u $scripts_environment_dir/build.py -i "${{ github.event.inputs.asset_dirs || env.default_asset_dirs }}" -a $asset_config_filename -o '${{ runner.temp }}'/$built_dir -l '${{ runner.temp }}'/$build_logs_artifact_name -P -g $resource_group -r $container_registry -T 'python -V' -c "${{ steps.changed-files.outputs.all_modified_files }}" -U "https://github.com/aquasecurity/trivy/releases/download/v0.42.1/trivy_0.42.1_Linux-64bit.deb"
```

File: scripts/azureml-assets/azureml/assets/environment/build.py [hunk 4]
```diff
@@ -62,6 +64,16 @@ def create_acr_task(image_name: str,
             'id': 'test',
             'cmd': f"$Registry/{image_name} {test_command}"
         })
+        task['steps'].append({
```
Comment (elesel) [inline scripts/azureml-assets/azureml/assets/environment/build.py:67]: If `trivy_url` is optional, you should check for `trivy_url is not None` before appending a scanning step. Maybe just use the URL to switch scanning on or off? If it's specified, then do vuln scanning. If not, don't?

**Review by elesel — APPROVED** [review]

**Review by elesel — DISMISSED** [review]

File: assets/training/general/environments/sklearn-1.0/context/Dockerfile [hunk 5]
```diff
@@ -16,3 +16,4 @@ RUN conda env create -p $CONDA_PREFIX -f conda_dependencies.yaml -q && \
     conda run -p $CONDA_PREFIX pip cache purge && \
     conda clean -a -y
 
+RUN echo "Hello World"
```
Comment (elesel) [inline assets/training/general/environments/sklearn-1.0/context/Dockerfile:19]: This still needs to be removed.

**Review by elesel — APPROVED** [review]

File: scripts/azureml-assets/azureml/assets/environment/build.py [hunk 6]
```diff
@@ -63,6 +65,18 @@ def create_acr_task(image_name: str,
             'cmd': f"$Registry/{image_name} {test_command}"
         })
 
+        if trivy_url is not None:
+            task['steps'].append({
+                'id': 'scan',
+                'cmd': (
+                        f"$Registry/{image_name} "
+                        f"wget -O trivy.deb {trivy_url} && "
```
Comment (elesel) [inline scripts/azureml-assets/azureml/assets/environment/build.py:73]: This minimizes output:
```suggestion
                        f"curl -s -S -f -o trivy.deb {trivy_url} && "
```

**Review by elesel — APPROVED** [review]

**Pull request merged**